src/middlewared/middlewared/plugins/nvidia.py

```diff
@@ -0,0 +1,130 @@
+import contextlib
+import os
+import shutil
+import subprocess
+import tempfile
+
+import requests
+
+from middlewared.service import job, Service
+from middlewared.utils.gpu import get_gpus
+from middlewared.utils.rootfs import ReadonlyRootfsManager
+
+HEADERS = {"User-Agent": "curl/7.88.1"}
+
+
+class NvidiaService(Service):
+    class Config:
+        private = True
+
+    def present(self):
+        adv_config = self.middleware.call_sync("system.advanced.config")
+
+        for gpu in get_gpus():
+            if gpu["addr"]["pci_slot"] in adv_config["isolated_gpu_pci_ids"]:
+                continue
+
+            if gpu["vendor"] == "NVIDIA":
+                return True
+
+        return False
+
+    def installed(self):
+        with open("/proc/modules") as f:
+            lines = f.readlines()
+
+        for line in lines:
+            if line.split()[0] == "nvidia":
+                return True
+
+        return False
+
+    @job(lock="nvidia", description=lambda *args: "Installing NVIDIA drivers")
+    def install(self, job, start_docker=None):
+        self.middleware.call_sync("network.general.will_perform_activity", "catalog")
+
+        with ReadonlyRootfsManager() as rrm:
```

Comment:
Before we enter this context manager, should we first check `self.installed` and return early?

Reply:
Oh, I didn't see a few lines up. Derp.